staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -450,7 +450,8 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 func (r *Reflector) syncWith(items []runtime.Object, resourceVersion string) error {
 	found := make([]interface{}, 0, len(items))
 	for _, item := range items {
-		found = append(found, item)
+		copyItem := item.DeepCopyObject()
```

Comment:
This appears to change the contract of this method by introducing a new deepcopy.  I'm surprised this is more efficient and I'm wary of making a change at this level.

Reply:
I'm very skeptical this is more efficient in all cases. I can imagine it would break up the memory allocated to a single LIST decode so that it would not be retained until all individual items from that list were released, at the cost of making the processing of that initial LIST much more expensive (I'd guess more than 2x the cost, since many small allocations are being done and many deep copies are being done)

Reply:
And that's just a memory alloc assessment.  I would expect overall time spent to increase in general over using a reference to the original data.

Reply:
+1 for David - I'm against this change

Reply:
Hi,all,other people encountered the same problem #102718

Reply:
I commented on that issue too - I don't think there is a bug here. It all depends what you do with the data, and that may lead to leak, not the way reflector is working.

Reply:
> This appears to change the contract of this method by introducing a new deepcopy. I'm surprised this is more efficient and I'm wary of making a change at this level.

Directly referencing the array will cause the underlying array to not be GC, and using `reflect` in `meta.ExtractList` to make a similar call, resulting in direct reference to the elements in the underlying array.
It hinders the work of GC. The specific call path is:

- Get the data list https://github.com/kubernetes/kubernetes/blob/b766721332da82f8a3eaa7dd3f131713d291f288/staging/src/k8s.io/client-go/tools/cache/reflector.go#L311
- Parse the item in the list, https://github.com/kubernetes/kubernetes/blob/b766721332da82f8a3eaa7dd3f131713d291f288/staging/src/k8s.io/client-go/tools/cache/reflector.go#L348
- **Directly ref the data in the list.** https://github.com/kubernetes/kubernetes/blob/b766721332da82f8a3eaa7dd3f131713d291f288/staging/src/k8s.io/apimachinery/pkg/api/meta/help.go#L180